xonsh/lazyasd.py

```diff
@@ -166,3 +174,106 @@ def __bool__(self):
         else:
             res = self._result
         return res
+
+#
+# Background module loaders
+#
+
+class BackgroundModuleProxy(types.ModuleType):
+    """Proxy object for modules loaded in the background that block attribute
+    access until the module is loaded..
+    """
+
+    def __init__(self, modname):
+        self.__dct__ = {
+            'loaded': False,
+            'modname': modname,
+            }
+
+    def __getattribute__(self, name):
+        passthrough = frozenset({'__dct__','__class__', '__spec__'})
+        if name in passthrough:
+            return super().__getattribute__(name)
+        dct = self.__dct__
+        modname = dct['modname']
+        if dct['loaded']:
+            mod = sys.modules[modname]
+        else:
+            delay_types = (BackgroundModuleProxy, type(None))
+            while isinstance(sys.modules.get(modname, None), delay_types):
+                time.sleep(0.001)
+            mod = sys.modules[modname]
+            dct['loaded'] = True
+        return getattr(mod, name)
+
+
+class BackgroundModuleLoader(threading.Thread):
+    """Thread to load modules in the background."""
+
+    def __init__(self, name, package, replacements, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.daemon = True
+        self.name = name
+        self.package = package
+        self.replacements = replacements
+        self.start()
+
+    def run(self):
+        # wait for other modules to stop being imported
+        i = 0
+        last = -6
+        hist = [-5, -4, -3, -2, -1]
+        while not all(last == x for x in hist):
+            time.sleep(0.001)
+            last = hist[i%5] = len(sys.modules)
+            i += 1
```

Comment:
What does this do. It is not immediately clear.  Does it check that the number modules haven't changed for 0.005 seconds? Then assume that all modules have loaded?